### PR TITLE
[kube-prometheus-stack] add serviceMonitor.additionalEndpoints

### LINKED ISSUE #3499: [kube-prometheus-stack] add `serviceMonitor.additionalEndpoints` field
### Is your feature request related to a problem ?

Hello.  First of all I just want to share my appreciation for everyone that's contributed to this project!  I can't state enough how much of a time save it is for everyone.  Amazing project, and also amazing keeping everything so tightly integrated and laid out well.  Very impressive.

I have added oauth2-proxy as a sidecar to my prometheus and alertmanager pods, and configured the ingress to point to this sidecar.  I have no issues with this setup - it is all nicely laid out in the comments of the values file.  However, oauth2-proxy supports exposing prometheus metrics on another port (so as to not conflict with the proxy port) and the only way to tell prometheus to scrape this new port is to create a new serviceMonitor.  Seeing as there already exists a serviceMonitor that points to this pod, I think a cleaner solution would be to add a new endpoint inside this serviceMonitor.

If this design is implemented, the same feature could be used for scraping any prometheus sidecar, not just oauth proxy.

I am planning to work on this feature request.  I just thought I should check here first to see if any the maintainers have thoughts before I start writing.

Thank you.

### Describe the solution you'd like.

I think the ideal solution would look like below.

```yaml
prometheus:
  serviceMonitor:
    additionalEndpoints:
      - path: /metrics
        port: oauth-metrics
alertmanager:
  serviceMonitor:
    additionalEndpoints:
      - path: /metrics
        port: oauth-metrics
```

### Describe alternatives you've considered.

currently what I have implemented is below.  As you can see it is much more verbose and redundant than appending onto existing serviceMonitor.

```yaml
prometheus:
  additionalServiceMonitors:
    # prometheus oauth-metrics
    - name: prometheus-prometheus-oauth
      selector:
        matchLabels:
          app: kube-prometheus-stack-prometheus
          release: prometheus
          self-monitor: 'true'
      endpoints:
        - path: /metrics
          port: oauth-metrics
    # alertmanager oauth-metrics
    - name: prometheus-alertmanager-oauth
      selector:
        matchLabels:
          app: kube-prometheus-stack-alertmanager
          release: prometheus
          self-monitor: 'true'
      endpoints:
        - path: /metrics
          port: oauth-metrics
```

### Additional context.

I am including a more full spec of my implementation so you can grasp it better incase I haven't explained it well.  I am including only the configuration for Prometheus for sake of brevity.

```yaml
prometheus:
  service:
    additionalPorts:
      - name: oauth-proxy
        port: 8081
        targetPort: 8081
      - name: oauth-metrics
        port: 8082
        targetPort: 8082
  ingress:
    enabled: true
    servicePort: 8081
  extraSecret:
    name: prometheus-oauth-config
    data:
      <REDACTED>
  prometheusSpec:
    containers:
      - name: oauth-proxy
        image: quay.io/oauth2-proxy/oauth2-proxy:v7.4.0
        args:
          - --upstream=http://127.0.0.1:9090
          - --http-address=0.0.0.0:8081
          - <more args>
          - --metrics-address=0.0.0.0:8082
        envFrom:
          - secretRef:
              name: prometheus-oauth-config
              optional: false
        ports:
          - containerPort: 8081
            name: oauth-proxy
            protocol: TCP
          - containerPort: 8082
            name: oauth-metrics
            protocol: TCP
  additionalServiceMonitors:
    # prometheus oauth-metrics
    - name: prometheus-prometheus-oauth
      selector:
        matchLabels:
          app: kube-prometheus-stack-prometheus
          release: prometheus
          self-monitor: 'true'
      endpoints:
        - path: /metrics
          port: oauth-metrics
```

### PR DESCRIPTION
**Pull request by TheRealNoob**
#### What this PR does / why we need it

Adds support for additional Endpoints in built-in serviceMonitors.  When adding an oauth-sidecar with `--metrics-address`, in order to add monitoring for it would require creating a 2nd ServiceMonitor.  This adds support for adding it as an endpoint on the existing ServiceMonitor.

```yaml
# desired spec
prometheus:
  serviceMonitor:
    additionalEndpoints:
      - path: /metrics
        port: oauth-metrics

# current spec
prometheus:
  additionalServiceMonitors:
    - name: prometheus-prometheus-oauth
      selector:
        matchLabels:
          app: kube-prometheus-stack-prometheus
          release: prometheus
      endpoints:
        - path: /metrics
          port: oauth-metrics
```

#### Which issue this PR fixes

- fixes #3499

#### Special notes for your reviewer

I also included several very small (related) changes in values.yaml.  Mostly comments.

#### Checklist

- [x] [DCO](https://github.com/prometheus-community/helm-charts/blob/main/CONTRIBUTING.md#sign-off-your-work) signed
- [x] Chart Version bumped
- [x] Title of the PR starts with chart name (e.g. `[prometheus-couchdb-exporter]`)

### CODE REVIEW

**Comment by QuentinBisson** [comment]
@TheRealNoob can you sign your commits?

**Comment by TheRealNoob** [comment]
@QuentinBisson Thank you for the msg.  Unfortunately I missed it in my email and it managed to slip my mind.  I've now updated it and am ready for review.  Couple talking points

* In the process of merging from master I saw that someone added a new endpoint to the same ServiceMonitor that this PR looks to change.  Previously I did a simpler approach of templating the passed list, but this time I changed it to look more similar to the other two endpoints.
* I didn't add any unittests.  Let me know if you need them.  But from the looks of it very little is tested using that approach.
* I've not used Thanos from this helm chart so my adding `additionalEndpoints` to it was a bit of a shot in the dark.  Let me know if it doesn't make sense of it I missed adding it to something more.

**Comment by TheRealNoob** [comment]
I think following semver this chart version should be bumped to 51.9.0 not 51.8.1 -- let me know.

File: charts/kube-prometheus-stack/templates/alertmanager/servicemonitor.yaml [hunk 1]
```diff
@@ -36,17 +36,15 @@ spec:
...
-{{- if .Values.alertmanager.serviceMonitor.metricRelabelings }}
-    metricRelabelings:
-{{ tpl (toYaml .Values.alertmanager.serviceMonitor.metricRelabelings | indent 6) . }}
-{{- end }}
-{{- if .Values.alertmanager.serviceMonitor.relabelings }}
-    relabelings:
-{{ toYaml .Values.alertmanager.serviceMonitor.relabelings | indent 6 }}
-{{- end }}
+    {{- if .Values.alertmanager.serviceMonitor.metricRelabelings }}
+    metricRelabelings: {{- tpl (toYaml .Values.alertmanager.serviceMonitor.metricRelabelings | nindent 6) . }}
```
Comment (QuentinBisson) [inline charts/kube-prometheus-stack/templates/alertmanager/servicemonitor.yaml:43]: Why would this even require nindent?

File: charts/kube-prometheus-stack/templates/alertmanager/servicemonitor.yaml [hunk 2]
```diff
@@ -36,17 +36,15 @@ spec:
...
+    {{- end }}
+    {{- if .Values.alertmanager.serviceMonitor.relabelings }}
+    relabelings: {{- toYaml .Values.alertmanager.serviceMonitor.relabelings | nindent 6 }}
```
Comment (QuentinBisson) [inline charts/kube-prometheus-stack/templates/alertmanager/servicemonitor.yaml:46]: Same remark

**Comment by QuentinBisson** [comment]
@TheRealNoob Could you fix the linting errors, check my comments and bump the version to 51.9.0? Then we should be good to go :)

File: charts/kube-prometheus-stack/templates/alertmanager/servicemonitor.yaml (hunk 1 above)
Comment (TheRealNoob) [reply charts/kube-prometheus-stack/templates/alertmanager/servicemonitor.yaml:43]: My purpose in changing this was just to make the code more readable.  The outcome should be identical.  I've just run a `helm template` on my branch as well the main branch (with a values file that defines both relabelings and metricsRelabelings) and and the outcomes are the identical.

The [documentation](https://github.com/prometheus-operator/prometheus-operator/blob/main/Documentation/api.md#endpoint) confirms both are expected to be a `[]RelabelConfig` list, so nindent sounds right to me.

**Comment by TheRealNoob** [comment]
This DCO thing is very annoying....

**Review by QuentinBisson — APPROVED** [review]

**Comment by QuentinBisson** [comment]
Thank you @TheRealNoob 🎉

**Pull request merged**